binder/src/main/java/io/grpc/binder/internal/PendingAuthListener.java

```diff
@@ -0,0 +1,97 @@
+package io.grpc.binder.internal;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+
+import java.util.concurrent.Executor;
+
+/**
+ * A {@link ServerCall.Listener} that can be returned by a {@link io.grpc.ServerInterceptor} to
+ * asynchronously advance the gRPC pending resolving a possibly asynchronous security policy check.
+ */
+final class PendingAuthListener<ReqT, RespT> extends ServerCall.Listener<ReqT> {
+
+    private final ListenableFuture<Listener<ReqT>> authStatusFuture;
+    private final Executor sequentialExecutor;
+    private final ObjectPool<? extends Executor> executorPool;
+    private final Executor executor;
+
+    /**
+     * @param authStatusFuture a ListenableFuture holding the result status of the authorization
+     *                         policy from a {@link io.grpc.binder.SecurityPolicy} or a
+     *                         {@link io.grpc.binder.AsyncSecurityPolicy}. The call only progresses
+     *                         if {@link Status#isOk()} is true.
+     * @param executorPool     a pool that can provide at least one Executor under which the result
+     *                         of {@code authStatusFuture} can be handled, progressing the gRPC
+     *                         stages.
+     * @param call             the 'call' parameter from {@link io.grpc.ServerInterceptor}
+     * @param headers          the 'headers' parameter from {@link io.grpc.ServerInterceptor}
+     * @param next             the 'next' parameter from {@link io.grpc.ServerInterceptor}
+     */
+    PendingAuthListener(
+            ListenableFuture<Status> authStatusFuture,
+            ObjectPool<? extends Executor> executorPool,
+            ServerCall<ReqT, RespT> call, Metadata headers,
+            ServerCallHandler<ReqT, RespT> next) {
+        this.executorPool = executorPool;
+        this.executor = executorPool.getObject();
+        this.authStatusFuture = Futures.transform(authStatusFuture, authStatus -> {
+            if (authStatus.isOk()) {
+                return next.startCall(call, headers);
+            }
+            call.close(authStatus, new Metadata());
+            throw new IllegalStateException("Auth failed", authStatus.asException());
+        }, executor);
+        this.sequentialExecutor = MoreExecutors.newSequentialExecutor(executor);
+    }
+
+    @Override
+    public void onCancel() {
+        ListenableFuture<?> unused = Futures.transform(authStatusFuture, delegate -> {
```

Comment:
Sorry to barge in but I believe this application of Futures.transform() may not forward ServerCall.Listener events to your delegate in the right order when authStatusFuture finally completes.  `sequentialExecutor` will certainly run your transformations in the order they are passed to execute(), but the order in which that happens upon completion is not guaranteed. The Futures.transform() javadoc doesn't guarantee it and, if you look at the current implementation, it boils down to a call to ListenableFuture#addListener which explicitly does not guarantee any particular listener execution order (https://guava.dev/releases/23.1-android/api/docs/com/google/common/util/concurrent/ListenableFuture.html#addListener-java.lang.Runnable-java.util.concurrent.Executor- )

Have a look at cs/symbol:DelayedServerCallListener for a class that might help you here.

Reply:
Seconding this suggestion.

I prefer the approach used by DelayedServerCallListener because it's also more "low level" which is appropriate when we're dealing with code potentially used by every RPC.